lxd/db_storage_pools.go

```diff
@@ -0,0 +1,520 @@
+package main
+
+import (
+	"database/sql"
+
+	_ "github.com/mattn/go-sqlite3"
+
+	"github.com/lxc/lxd/shared"
+	"github.com/lxc/lxd/shared/api"
+)
+
+// Get all storage pools.
+func dbStoragePools(db *sql.DB) ([]string, error) {
+	var name string
+	query := "SELECT name FROM storage_pools"
+	inargs := []interface{}{}
+	outargs := []interface{}{name}
+
+	result, err := dbQueryScan(db, query, inargs, outargs)
+	if err != nil {
+		return []string{}, err
+	}
+
+	if len(result) == 0 {
+		return []string{}, NoSuchObjectError
+	}
+
+	pools := []string{}
+	for _, r := range result {
+		pools = append(pools, r[0].(string))
+	}
+
+	return pools, nil
+}
+
+// Get the names of all storage volumes attached to a given storage pool.
+func dbStoragePoolsGetDrivers(db *sql.DB) ([]string, error) {
+	var poolDriver string
+	query := "SELECT DISTINCT driver FROM storage_pools"
+	inargs := []interface{}{}
+	outargs := []interface{}{poolDriver}
+
+	result, err := dbQueryScan(db, query, inargs, outargs)
+	if err != nil {
+		return []string{}, err
+	}
+
+	if len(result) == 0 {
+		return []string{}, NoSuchObjectError
+	}
+
+	drivers := []string{}
+	for _, driver := range result {
+		drivers = append(drivers, driver[0].(string))
+	}
+
+	return drivers, nil
+}
+
+// Get id of a single storage pool.
+func dbStoragePoolGetID(db *sql.DB, poolName string) (int64, error) {
+	poolID := int64(-1)
+	query := "SELECT id FROM storage_pools WHERE name=?"
+	inargs := []interface{}{poolName}
+	outargs := []interface{}{&poolID}
+
+	err := dbQueryRowScan(db, query, inargs, outargs)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return -1, NoSuchObjectError
+		}
+	}
+
+	return poolID, nil
+}
+
+// Get a single storage pool.
+func dbStoragePoolGet(db *sql.DB, poolName string) (int64, *api.StoragePool, error) {
+	var poolDriver string
+	poolID := int64(-1)
+	query := "SELECT id, driver FROM storage_pools WHERE name=?"
+	inargs := []interface{}{poolName}
+	outargs := []interface{}{&poolID, &poolDriver}
+
+	err := dbQueryRowScan(db, query, inargs, outargs)
+	if err != nil {
+		if err == sql.ErrNoRows {
```

Comment:
That one should indeed be interpreted properly by SmartError, so just letting it through should work.

There is also a bigger issue here which is that if dbQueryRowScan returns an error other than ErrNoRows, then we'll just continue as if nothing happened.